cluster/engine.go

```diff
@@ -0,0 +1,527 @@
+func (n *Engine) Addr() string {
+	return n.addr
+}
+
+// Name of the docker engine
+func (n *Engine) Name() string {
+	return n.name
+}
+
+// Labels of the docker engine
+func (n *Engine) Labels() map[string]string {
+	return n.labels
+}
+
+// Connect will initialize a connection to the Docker daemon running on the
+// host, gather machine specs (memory, cpu, ...) and monitor state changes.
+func (n *Engine) Connect(config *tls.Config) error {
+	host, _, err := net.SplitHostPort(n.addr)
+	if err != nil {
+		return err
+	}
+
+	addr, err := net.ResolveIPAddr("ip4", host)
+	if err != nil {
+		return err
+	}
+	n.ip = addr.IP.String()
+
+	c, err := dockerclient.NewDockerClientTimeout("tcp://"+n.addr, config, time.Duration(requestTimeout))
+	if err != nil {
+		return err
+	}
+
+	return n.ConnectClient(c)
+}
+
+// ConnectClient get specs of the engine (/info) and then refresh containers and images
+func (n *Engine) ConnectClient(client dockerclient.Client) error {
+	n.Client = client
+
+	// Fetch the engine labels.
+	if err := n.updateSpecs(); err != nil {
+		n.Client = nil
+		return err
+	}
+
+	// Force a state update before returning.
+	if err := n.refreshContainers(true); err != nil {
+		n.Client = nil
+		return err
+	}
+
+	if err := n.refreshImages(); err != nil {
+		n.Client = nil
+		return err
+	}
+
+	// Start the update loop.
+	go n.refreshLoop()
+
+	// Start monitoring events from the node.
+	n.Client.StartMonitorEvents(n.handler, nil)
+	n.emitEvent("node_connect")
+
+	return nil
+}
+
+// IsConnected returns true if the engine is connected to a remote docker API
+func (n *Engine) IsConnected() bool {
+	return n.Client != nil
+}
+
+// IsHealthy returns true if the engine is healthy
+func (n *Engine) IsHealthy() bool {
+	return n.healthy
+}
+
+// Gather node specs (CPU, memory, constraints, ...).
+func (n *Engine) updateSpecs() error {
+	info, err := n.Client.Info()
+	if err != nil {
+		return err
+	}
+	// Older versions of Docker don't expose the ID field and are not supported
+	// by Swarm.  Catch the error ASAP and refuse to connect.
+	if len(info.ID) == 0 {
+		return fmt.Errorf("node %s is running an unsupported version of Docker Engine. Please upgrade.", n.addr)
+	}
+	n.id = info.ID
+	n.name = info.Name
+	n.Cpus = info.NCPU
+	n.Memory = info.MemTotal
+	n.labels = map[string]string{
+		"storagedriver":   info.Driver,
+		"executiondriver": info.ExecutionDriver,
+		"kernelversion":   info.KernelVersion,
+		"operatingsystem": info.OperatingSystem,
+	}
+	for _, label := range info.Labels {
+		kv := strings.SplitN(label, "=", 2)
+		n.labels[kv[0]] = kv[1]
+	}
+	return nil
+}
+
+// RemoveImage deletes an image from the node.
+func (n *Engine) RemoveImage(image *Image) ([]*dockerclient.ImageDelete, error) {
+	return n.Client.RemoveImage(image.Id)
+}
+
+// Refresh the list of images on the node.
+func (n *Engine) refreshImages() error {
+	images, err := n.Client.ListImages()
+	if err != nil {
+		return err
+	}
+	n.Lock()
+	n.images = nil
```

Comment:
Seems if `n.images` is not used anywhere at that time it's better to reuse allocated memory:
```
n.images = n.images[:0]
```